### ISSUE TITLE
`getInitialProps` in `_document` does not produce server-side `styled-components` styles

### ISSUE THREAD

**Issue by kachkaev** [vercel/next.js#35758]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release
   (12.1.3 is the latest release at the time of writing)

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:35:58 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.14.0
      npm: 8.5.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.1.3
      react: 18.0.0
      react-dom: 18.0.0

### What browser are you using? (if relevant)

N/A

### How are you deploying your application? (if relevant)

N/A

### Describe the Bug

This bug is very similar to what has already been described in https://github.com/vercel/next.js/issues/35711 and fixed in https://github.com/vercel/next.js/pull/35736. Upgrading Next from 12.1.1 to 12.1.3 breaks SSR in `styled-components` when using React 18. Initial styles are missing in page source, as if `getInitialProps` in `_document` have not been called.

```
...<style data-styled="" data-styled-version="5.3.5">body{margin:0;padding:0;box-sizing:border-box;}/*!sc*/
data-styled.g1[id="sc-global-fvFGgF1"]{content:"sc-global-fvFGgF1,"}/*!sc*/
.hAcELG{font-size:50px;color:#0070f3;}/*!sc*/
data-styled.g2[id="pages__Title-sc-eaab758a-0"]{content:"hAcELG,"}/*!sc*/
</style></head>...
```

This may be to do with hydration too (not sure).

### Expected Behavior

I expect no initial flickering in styles when opening a server-side-rendered page.

### To Reproduce

1.  `yarn create next-app --example with-styled-components with-styled-components-app`

1.  Upgrade React to `18.0.0`

1.  Run `yarn dev`

1.  Open `view-source:http://localhost:3000`

1.  Observe that there is no style tag `<style data-styled="" data-styled-version="5.3.5">...</style>`. The whole page source fits one line.

1.  Downgrade Next from `12.1.3` to `12.1.2` and confirm the same behavior.

1.  Downgrade Next from `12.1.2` to `12.1.1`, open `view-source:http://localhost:3000` and observe the style tag. The source will span across five lines. This is expected.

The example is too small to observe the actual style flickering, but it is noticeable in a real Next.js app.

Relevant example code:

https://github.com/vercel/next.js/blob/9f9214abe58079e92cb42d50f537587d364e58f3/examples/with-styled-components/pages/_document.js#L5-L29

**Comment by ijjk** [comment]
Hi, this has been updated in the latest version of Next.js `v12.1.4-canary.1`, please update and give it a try!

**Comment by kachkaev** [comment]
Thank you for releasing the fix so quickly! I upgraded to `12.1.4` and all worked! 🎉
